### PR TITLE
Handle logout in SPA

### LINKED ISSUE #675: Handle logout in SPA
Following up on #637, the user should be able to logout from the provisioning app. A `/logout` endpoint is already available.

### PR DESCRIPTION
**Pull request by netzartist**
Fixes #675

### CODE REVIEW

File: client/src/app/app.component.ts [hunk 1]
```diff
@@ -99,10 +105,14 @@ export class AppComponent implements OnInit {
     this.projectService
       .getAllProjects()
       .pipe(
-        catchError(() => {
+        catchError((response: HttpErrorResponse) => {
+          if (response.error.status === 401) {
```
Comment (cschweikert) [inline client/src/app/app.component.ts:109]: When is the interceptor jumping in? Isn't it intercepting even before this line is executed?

File: client/src/app/app.component.ts [hunk 2]
```diff
@@ -99,10 +105,14 @@ export class AppComponent implements OnInit {
...
+            // interceptor takes control and redirects to login
+            return EMPTY;
+          }
+          // show generic error page
           this.isLoading = false;
           this.isError = true;
-          return EMPTY;
```
Comment (cschweikert) [inline client/src/app/app.component.ts:105]: Does it work without returning (another) observable?

Comment (netzartist) [reply client/src/app/app.component.ts:105]: No, returning an `EMPTY` observable is necessary ;)

File: client/src/app/app.component.ts (hunk 1 above)
Comment (netzartist) [reply client/src/app/app.component.ts]: Right, I was a bit on the wrong track due to a bug in the interceptor which is fixed now. In 401 cases the interceptor needs to return an `EMPTY` observable in order to let the app component render the login page through the router outlet in this case.

File: client/src/app/app.component.ts [hunk 3]
```diff
@@ -55,24 +56,8 @@ export class AppComponent implements OnInit {
...
-  }
-
-  showLogoutButton() {
-    this.router.events.subscribe(event => {
-      if (event instanceof NavigationStart && (event.url === 'login' || event.url === 'logout')) {
-        return false;
-      }
-      return true;
-    });
+    this.checkShowLogoutButton();
```
Comment (cschweikert) [inline client/src/app/app.component.ts:60]: Maybe just a minor thing: a few lines above we are already looking at ``router.events``. Maybe the code can be joined, e.g. in the following way:
```
this.router.events.subscribe(event => {
  if (event instanceof NavigationStart) {
    if (event.url === '/') {
      this.loadAllProjects();
    } else if (event.url === 'login' || event.url === 'logout') {
      this.showLogoutButton = false;
    }
    this.isLoading = false;
  }
});
```

BTW: I'm wondering, if ``this.isLoading = false;`` should not be skipped when calling ``this.loadAllProjects();``.

File: client/src/app/app.component.ts [hunk 4]
```diff
@@ -89,10 +74,29 @@ export class AppComponent implements OnInit {
...
+  private checkShowLogoutButton() {
+    return this.router.events.subscribe(event => {
+      if (event instanceof NavigationStart && (event.url === '/login' || event.url === '/logout')) {
+        this.showLogoutButton = false;
+      }
+    });
+  }
+
+  private setSsoMode() {
+    this.authenticationService.sso = true;
```
Comment (cschweikert) [inline client/src/app/app.component.ts:86]: just a thought: Would it make sense to move the whole sso detection (``this.activatedRoute.queryParams.subscribe(...)`` or maybe just the ``setSsoMode()`` function including the ``this.router.navigate(...)`` into the ``AuthenticationService``?

**Review by cschweikert — APPROVED** [review]

File: client/src/app/app.component.ts (hunk 3 above)
Comment (netzartist) [reply client/src/app/app.component.ts]: Good point! And `this.isLoading` should be in the `else if` part, as `this.loadAllProjects()` sets this on its own.

File: client/src/app/app.component.ts (hunk 4 above)
Comment (netzartist) [reply client/src/app/app.component.ts]: Oh yeah, that's a better separation of concerns.

File: client/src/app/app.component.ts [hunk 5]
```diff
@@ -40,35 +42,20 @@ export class AppComponent implements OnInit {
...
-        this.setSsoMode();
-      }
-    });
+    this.authenticationService.checkAndSetSsoMode();
 
     this.router.events.subscribe(event => {
       if (event instanceof NavigationStart) {
         if (event.url === '/') {
           this.loadAllProjects();
+        } else if (event instanceof NavigationStart && (event.url === '/login' || event.url === '/logout')) {
```
Comment (cschweikert) [inline client/src/app/app.component.ts:51]: ``event instanceof NavigationStart && `` can be skipped here. It is checked earlier

File: client/src/app/app.component.ts [hunk 6]
```diff
@@ -40,35 +42,20 @@ export class AppComponent implements OnInit {
   }
 
   ngOnInit() {
-    this.activatedRoute.queryParams.subscribe(params => {
```
Comment (cschweikert) [inline client/src/app/app.component.ts:43]: ``this.activatedRoute`` can also be removed from the constructor

File: client/src/app/app.component.ts (hunk 5 above)
Comment (netzartist) [reply client/src/app/app.component.ts]: Done

File: client/src/app/app.component.ts (hunk 6 above)
Comment (netzartist) [reply client/src/app/app.component.ts:43]: Done

**Review by cschweikert — APPROVED** [review]

**Pull request merged**